### PR TITLE
Logic cleanups for `basic_string`

### PR DESCRIPTION
**Pull request by achabense**
(Updated)
1. Makes `string(const T*)` and `string(const T*, const Alloc&)` adjacent. They work as a whole to resolve to `string(const T*, const Alloc&={})`.
2. Rectify usages of small/large string mode. These changes will eliminate all direct comparisons with _BUF_SIZE in basic_string.
- Introduces `_SMALL_STRING_CAPACITY`, which equals `_BUF_SIZE - 1`, and replace every `_BUF_SIZE - 1` in `basic_string` with it.
- Introduces `_Entails_large_string(_Size/_Cap)`, which decides whether newly planned size/capacity ***requires*** large string mode.
- Refines `_Large_string_engaged` logic, and fixes a relevant misuse in `operator=(const string&)` (which will break invariant, and should be `_Entails_large_string(_Mysize)` instead).
- Introduces `_LEAST_ALLOCATION_SIZE` for `_Move_assign_from_buffer` and `_Release_to_buffer`.
- These changes will eliminate all direct comparisons with `_BUF_SIZE` in `basic_string`.
3. Rectify the relation between capacity and allocation size.
- Introduces `_Allocate_for_capacity`, which does stl check and lifetime activation, and takes capacity as input. This will eliminate all `(al.allocate(res+1) ... _Start_element_lifetimes)` pattern.
- Introduces `_Deallocate_for_capacity` counterpart.
- These two changes will remove all direct usages of `al.deallocate`, and only one wild `al.allocate`(without lifetime startup) is left in the interaction part with `basic_stringbuf`.

2 and 3 will make it easier to keep the following invariant: 
`capacity always== _SMALL_STRING_CAPACITY in small mode && capacity always> _SMALL_STRING_CAPACITY in large mode`

### CODE REVIEW

**Comment by achabense** [comment]
Some added names/comments might be terrible in the original pr 😳...

File: stl/inc/xstring
Comment (achabense) [inline stl/inc/xstring:2620]: I'm not too confident about this one. Is passing `pointer` by value ok here?

**Comment by achabense** [comment]
looks like some timeout...

File: stl/inc/xstring
Comment (achabense) [inline stl/inc/xstring:3180]: The logic of `_(De)allocate_from_al` is self-consistent. However we still need to care about '\0' when it comes to `_Mysize`. (e.g. `_Traits::copy(_Unfancy(_New_ptr), _Unfancy(_Right._Mypair._Myval2._Bx._Ptr), _New_size + 1);`)... I think we should at least do some renamings to make it clear that `_(De)allocate_from_al` can do the right thing with **`capacity`**.

File: stl/inc/xstring
Comment (frederick-vs-ja) [reply stl/inc/xstring:2620]: > I'm not too confident about this one. Is passing `pointer` by value ok here?

IMO yes. If someone provides a `pointer` type which is not OK for passing by value, that the allocator won't meet the _Cpp17Allocator_ named requirements and thus the behavior is undefined.

File: stl/inc/xstring
Comment (fsb4000) [inline stl/inc/xstring:2372]: >wether

I believe it is a typo and it should be "whether"

Comment (achabense) [reply stl/inc/xstring:2372]: Yes 😳, and the whole pr might need some rewordings... I'm pretty uncomfortable with my English...

**Comment by achabense** [comment]
When replacing `_BUF_SIZE - 1` with `_SMALL_STRING_CAPACITY`, I noticed an exception here:
https://github.com/microsoft/STL/blob/40640c6df9754ea031f8df7c720b89714d19df97/stl/inc/xstring#L3170
This is the only static call that starts with `0` instead of `_BUF_SIZE - 1`. The second parameter is called `_Old`, so this does not look like a good thing.

**Comment by achabense** [comment]
I've made a new push to try to make the behavior in `operator=(const basic_string&::if (_Al != _Right_al)` not breaking invariant(not allocating buffer smaller than `BUF_SIZE`)...

File: stl/inc/xstring
Comment (achabense) [reply stl/inc/xstring]: fixed in a new push

File: stl/inc/xstring
Comment (achabense) [reply stl/inc/xstring:2620]: My concern is that, `pointer` looks not necessarily trivially copyable, so it's possible that this call will not be an exact replacement to direct `_Al.deallocate` call in some rare cases.

File: stl/inc/xstring
Comment (achabense) [inline stl/inc/xstring:3169]: I think dropping some slight efficiency gain from `_Memcpy_val_from` is ok here.
Otherwise we have to deal with three cases:
right is in small mode(res&size) -> can do `_Memcpy_val_from` in non-constexpr mode.
right is in large(res) mode, but is small(size)
right is in large mode(res) and is large(size)

File: stl/inc/xstring
Comment (achabense) [inline stl/inc/xstring:3170]: This part was broken and can cause allocation smaller than `BUF_SIZE`. (e.g. consider when `large_string_engaged()`, but `size() == 0`)

File: stl/inc/xstring
Comment (achabense) [inline stl/inc/xstring:4733]: Should we add internal checks in these cases(`_Old_capacity>=_SMALL_STRING_CAPACITY`, like in `_Large_string_engaged`)?

File: stl/inc/xstring
Comment (achabense) [inline stl/inc/xstring:2246]: Some added comments in this pr look awkward and need some rewordings😳...

Comment (frederick-vs-ja) [reply stl/inc/xstring:2246]: I think we may say "determines whether large mode is used for such capacity". Also I think the name of this function should be changed to `_Is_large_capacity`.

Comment (achabense) [reply stl/inc/xstring:2246]: I believe the name is bad too. However, renaming is not enough, as this function is widely referred to in comments, and there are a lot of other usages of "large" in the class(e.g. `_My_large`, `_Is_large`, `_Right_large`, `//... large mode...`).

I think The usage of `large` is generally distracting, as most of them actually mean allocation. For long-time maintainability, we may need to remove this indirection of meaning from the whole class. However, I think we'd better put this work off -- firstly making an issue about main leftover problems in `basic_string` after this pr being settled, and then making separate prs to do the fix.

Some problems that may need further fix:
1. usage of `"large"`
2. coexistence of `"res"` and `"capacity"` for the same meaning
3. handling of `\0` when doing `copy`. Currently they are hard-code maintained like `traits::copy/move(...,_Size+1)`

File: stl/inc/xstring
Comment (achabense) [inline stl/inc/xstring:2379]: Are there benefits to make trivial internal methods `_CONSTEXPR20` instead of `constexpr`? I see `_Large_string_engaged` use `_CONSTEXPR20`.

**Comment by fsb4000** [comment]
> Are there benefits to make trivial internal methods _CONSTEXPR20 instead of constexpr? I see _Large_string_engaged use _CONSTEXPR20.

Since C++20, `std::string` has become a `constexpr` type. To ensure consistency, we use `_CONSTEXPR20` for all methods instead of mixing `_CONSTEXPR20`(for public methods) and `constexpr`(for internal methods).

**Comment by StephanTLavavej** [comment]
The test failures look like a few tests are legitimately hanging, although I have not yet attempted to reproduce this locally.

File: stl/inc/xstring
Comment (StephanTLavavej) [reply stl/inc/xstring]: "Large" is the antonym of "small" - I don't see a problem there. (My opinion shouldn't be taken as definitive - I have lived with this codebase for so long, weird things can be familiar to me.)

**Comment by StephanTLavavej** [comment]
There are indeed 5,387 test failures. `P2093R14_formatted_output` is the one that's hanging. (`--timeout=240` can be used to deal with that, if you've installed `psutil`.)

Full test logs from my local machine: [testing_x64.txt.zip](https://github.com/microsoft/STL/files/11923335/testing_x64.txt.zip)

Obtained via running all tests, then decoding the JSON to be human-readable:

```cmd
python tests\utils\stl-lit\stl-lit.py D:\GitHub\STL\tests\std D:\GitHub\STL\tests\tr1 D:\GitHub\STL\llvm-project\libcxx\test --order=random -o testing_x64.log
python D:\GitHub\STL\tools\scripts\print_failures.py testing_x64.log
```

(I didn't have psutil installed, so instead of using the timeout option, I just terminated the hanging test executables.)

Many failures, like `Dev11_0091392_string_erase_resize_perf`, look like:

```
D:\GitHub\STL\out\x64\out\inc\xstring(2247) : Assertion failed: STL internal check: _Strcap >= _SMALL_STRING_CAPACITY
```

There are others, like `P1206R7_string_replace_with_range` which begins with:

```
D:\GitHub\STL\out\x64\out\inc\xstring(2734): error C2064: term does not evaluate to a function taking 0 arguments
D:\GitHub\STL\out\x64\out\inc\xstring(2773): note: see reference to function template instantiation 'void std::basic_string<char,std::char_traits<char>,std::allocator<char>>::_Construct_from_iter<test::iterator<Category,Element,test::CanDifference::no,test::CanCompare::no,test::ProxyRef::yes,test::WrappedState::unwrapped>,test::sentinel<Element,test::WrappedState::unwrapped>,std::nullptr_t>(_Iter,const _Sent,_Size)' being compiled
        with
        [
            Category=std::input_iterator_tag,
            Element=const char,
            _Iter=test::iterator<std::input_iterator_tag,const char,test::CanDifference::no,test::CanCompare::no,test::ProxyRef::yes,test::WrappedState::unwrapped>,
            _Sent=test::sentinel<const char,test::WrappedState::unwrapped>,
            _Size=std::nullptr_t
        ]
```

File: stl/inc/xstring
Comment (StephanTLavavej) [inline stl/inc/xstring:2734]: The empty parens here are bogus and causing test failures.

**Review by StephanTLavavej — CHANGES_REQUESTED** [review]

File: stl/inc/xstring
Comment (achabense) [inline stl/inc/xstring:2302]: I realize that `_Myres` should be initialized to `_SMALL_STRING_CAPACITY`, as `_String_val` begins with SSO mode.
After this change, however, I'm getting `static_assert(test_to_string<63>()); ... constexpr evaluation hit maximum step limit` in [this test](https://github.com/llvm/llvm-project/blob/dc6c8b8d1e357acf6440824afaf6b6547b34aeeb/libcxx/test/std/utilities/template.bitset/bitset.members/to_string.pass.cpp#L118).

Comment (achabense) [reply stl/inc/xstring:2302]: If this change is valid, then some further changes are needed(e.g. removing redundant assignments like `_Myres = _SMALL_STRING_CAPACITY` in basic_string's constructors)

Comment (frederick-vs-ja) [reply stl/inc/xstring:2302]: > After this change, however, I'm getting `static_assert(test_to_string<63>()); ... constexpr evaluation hit maximum step limit` in [this test](https://github.com/llvm/llvm-project/blob/dc6c8b8d1e357acf6440824afaf6b6547b34aeeb/libcxx/test/std/utilities/template.bitset/bitset.members/to_string.pass.cpp#L118).

This is because of redundant checking. Do you think it make sense to check in `_Large_string_engaged` which is heavily called?

Comment (achabense) [reply stl/inc/xstring:2302]: I intended to use `_INTERNAL_CHECK` to express invariant here. I think it will not expand to live code in most cases.
However, yes, I'm uncertain whether this is good practice. As the function is only called in several places, it might be useless and just some noise. And due to early call it's introducing a stricter invariant: `_Myres` should >= `_SMALLSTRING_CAPACITY`, `from the beginning of lifetime`(this looks not too bad for me).

File: stl/inc/xstring
Comment (achabense) [reply stl/inc/xstring]: > I think we may say "determines whether large mode is used for such capacity"

adopted in new push

File: stl/inc/xstring
Comment (achabense) [reply stl/inc/xstring]: I decide to remove this check to pass those clang tests.

File: stl/inc/xstring
Comment (achabense) [inline stl/inc/xstring]: It was this check(added in this pr) that were causing lots of test failures, as
1. `_Myres` is initialized to `0` on construction.
2. In several places(like [`_Construct`](https://github.com/microsoft/STL/blob/2261f7edb760eb3fe0726187c818b796dc7ea798/stl/inc/xstring#L2610-L2612)), `_Large_string_engaged` is called early before assignment of `_Myres`.

After making `_Myres` initialized to `_SMALL_STRING_CAPACITY`, I made some checks passed, however there were some errors in clang checks due to `constexpr evaluation hit maximum step limit`.

Here I decide to remove this check to pass those clang tests. However, I'm uncertain wether this check is helpful. **Does the check make sense? Should I add it back?**

And I'm leaving `_Myres` initialized to `_SMALL_STRING_CAPACITY`, as `_String_val` begins with SSO mode. This change will make some assignments redundant: (Candidates: [(1)](https://github.com/microsoft/STL/blob/2261f7edb760eb3fe0726187c818b796dc7ea798/stl/inc/xstring#L2630) [(2)](https://github.com/microsoft/STL/blob/2261f7edb760eb3fe0726187c818b796dc7ea798/stl/inc/xstring#L2650) [(3)](https://github.com/microsoft/STL/blob/2261f7edb760eb3fe0726187c818b796dc7ea798/stl/inc/xstring#L2686); they are unchanged now)

File: stl/inc/xstring
Comment (achabense) [inline stl/inc/xstring:2303]: I'm leaving `_Myres` initialized to `_SMALL_STRING_CAPACITY`, as `_String_val` begins with SSO mode. This change will make some assignments redundant: (Candidates: [(1)](https://github.com/microsoft/STL/blob/2261f7edb760eb3fe0726187c818b796dc7ea798/stl/inc/xstring#L2630) [(2)](https://github.com/microsoft/STL/blob/2261f7edb760eb3fe0726187c818b796dc7ea798/stl/inc/xstring#L2650) [(3)](https://github.com/microsoft/STL/blob/2261f7edb760eb3fe0726187c818b796dc7ea798/stl/inc/xstring#L2686); they are unchanged now)

**Comment by StephanTLavavej** [comment]
I am generally concerned about the magnitude of these changes; there are a lot of transformations happening here, and having them all in a single PR (without well-structured fine-grained commits) makes the validity of each transformation difficult to review. I think it's reviewable but I'll need some time.

**Comment by achabense** [comment]
Do I need to reorganize commits and reupload this pr?

File: stl/inc/xstring
Comment (StephanTLavavej) [inline stl/inc/xstring:2249]: This is a philosophical point, but I strongly believe that function overloading should be used only when the programmer "doesn't care" which overload is selected. For example, overloading `to_string(short)` and `to_string(unsigned long long)` is totally fine, because the programmer has already chosen a data type, and now they want a `string`, and they don't care how it happens.

Here, there is a *major* difference between the `static` and non-`static` overloads of `_Large_string_engaged()`, which is dangerous. The `static` doesn't prevent mistakes, because `static` functions can be called with non-`static` syntax.

`_Entails_large_string()` has the exact same behavior, but a much better name (the different word order is critical); I believe we should use it uniformly.

File: stl/inc/xstring
Comment (StephanTLavavej) [inline stl/inc/xstring:2560]: This is a desirable change to avoid splitting the constructor implementation, but `basic_string(nullptr_t)` is ending up in the "wrong" place compared to its appearance in the Standard. We're far from consistent (*especially* in old code), but nowadays we try to match the Standard's declaration order unless there's a good reason otherwise.

**Comment by StephanTLavavej** [comment]
> I think I can group up these changes. Do I need to reorganize commits and reupload this pr?

If you have time, that would be great! I started doing that, but I ran out of time today (and will need to work on a Patch Tuesday toolset update tomorrow before I can resume reviewing this PR). I pushed a conflict-free merge with `main` plus a couple of changes for issues that I identified; please pick up these changes (assuming you agree with them, of course!).

You can see https://github.com/microsoft/STL/compare/main...StephanTLavavej:STL:fine-grained for an idea of what commits would be easiest to review. When each commit is focused on a single transformation whose correctness can be verified at each location, or is focused on a single fix, then they can be reviewed sequentially much easier without the risk of missing anything.

We usually discourage force-pushes for PRs that are under review, since GitHub doesn't make it easy to see what changed. In this case, because I've paused reviewing and the other maintainers are all busy, it's your call - you can either force-push this PR, or you can close this PR and open a new one. Either way, I recommend verifying that your commit restructuring isn't introducing unintentional changes - you can simply `git diff OLD_BRANCH NEW_BRANCH` to verify that they're ending up with the same content.

Thanks! :heart_eyes_cat: Your changes look like a great refactoring, I just want to be really careful about reviewing it given how complicated `string` is, and how absolutely critical its correctness is.

**Comment by achabense** [comment]
Ok, I will do that. I'm going to close this pr and make a new one.

**Pull request closed without merging**
